### PR TITLE
Minigame PR

### PR DESCRIPTION
**Pull request by kywch**
Created a PR so that the changes can be reviewed. 

* Created the minigame API and changed the config, so that each episode can run in different config while keeping the obs shape same
* Speed up the mask generation by reusing the np arrays and cythonizing a few functions
* Made the env API conform to the new pettingzoo
* Bumped up the deps (e.g., replaced gym with gymnasium, removed old renderer stuff)
* Refactored NPC, etc, so that the directory structure is much simpler.

Please feel free to make suggestions. Thank you!

### CODE REVIEW

File: nmmo/core/action.py [hunk 1]
```diff
@@ -48,7 +45,7 @@ def leaf():
   def N(cls, config):
     return len(cls.edges)
 
-  def deserialize(realm, entity, index, obs: Observation):
+  def deserialize(realm, entity, index: int, obs):
```
Comment (daveey) [inline nmmo/core/action.py]: why remove type hint?

File: nmmo/core/action.py [hunk 2]
```diff
@@ -176,7 +176,7 @@ def deserialize_fixed_arg(arg, index):
     val = min(index, len(arg.edges)-1)
     return arg.edges[val]
 
-  # if index is not int, it's probably already deserialized
+  # if index is not int, it"s probably already deserialized
```
Comment (daveey) [inline nmmo/core/action.py]: typo

File: nmmo/core/action.py [hunk 3]
```diff
@@ -243,26 +243,26 @@ def call(realm, entity, style, target):
...
       return None
 
-    #Check if self targeted
-    if entity.ent_id == target.ent_id:
+    #Check if self targeted or target already dead
+    if entity.ent_id == target.ent_id or not target.alive:
       return None
 
-    #Can't attack out of range
+    #Can"t attack out of range
```
Comment (daveey) [inline nmmo/core/action.py]: typo

File: nmmo/core/action.py [hunk 4]
```diff
@@ -243,26 +243,26 @@ def call(realm, entity, style, target):
...
+    entity.history.attack["style"] = style.__name__
     target.attacker = entity
     target.attacker_id.update(entity.ent_id)
 
     from nmmo.systems import combat
     dmg = combat.attack(realm, entity, target, style.skill)
 
-    if style.freeze and dmg > 0:
-      target.status.freeze.update(config.COMBAT_FREEZE_TIME)
+    # if style.freeze and dmg > 0:
```
Comment (daveey) [inline nmmo/core/action.py]: did you mean to leave this commented? if so, explain why

File: nmmo/core/action.py (hunk 1 above)
Comment (kywch) [reply nmmo/core/action.py]: I put back type hint where it's helpful, including here.

File: nmmo/core/action.py (hunk 2 above)
Comment (kywch) [reply nmmo/core/action.py]: fixed!

File: nmmo/core/action.py (hunk 3 above)
Comment (kywch) [reply nmmo/core/action.py]: fixed!

File: nmmo/core/action.py (hunk 4 above)
Comment (kywch) [reply nmmo/core/action.py]: Removed these completely, as there is no freeze attack.

**Comment by kywch** [comment]
Thank you for your review, David! I'll yolo-merge this one.

**Pull request merged**